File: python/63_Unique_Paths_II.py
#!/usr/bin/env python3

class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """

        if obstacleGrid is None or len(obstacleGrid) == 0:
            return 0
        if obstacleGrid[0] is None or len(obstacleGrid[0]) == 0:
            return 0

        m = len(obstacleGrid)
        n = len(obstacleGrid[0])

        col = [0] * n
        col[-1] = 1

        for i in range(m)[::-1]:
            for j in range(n)[::-1]:
                if obstacleGrid[i][j] == 1:
                    col[j] = 0
                elif j != n-1:
                    col[j] += col[j+1]

        return col[0]

s = Solution()
print(s.uniquePathsWithObstacles([
  [0,0,0],
  [0,1,0],
  [0,0,0]
]))

File: java/src/brand_new/_63_Unique_Paths_II.java
package brand_new;

public class _63_Unique_Paths_II {

	public int uniquePathsWithObstacles(int[][] obstacleGrid) {
		if(obstacleGrid.length == 0 || obstacleGrid[0].length == 0)
			return 0;
		int m = obstacleGrid.length;
		int n = obstacleGrid[0].length;
		
		int[][] matrix = new int[m][n];
		matrix[0][0] = 1;
		for(int i = 0; i < m; i++){
			for(int j = 0; j < n; j++){
				if(obstacleGrid[i][j] != 1){
					if(i != 0)
						matrix[i][j] += matrix[i-1][j];
					if(j != 0)
						matrix[i][j] += matrix[i][j-1];
				}
			}
		}
		return matrix[m-1][n-1];
	}

	public static void main(String[] args) {

	}

}
